.gitmodules

```diff
@@ -0,0 +1,9 @@
+[submodule "tests/integration/pip"]
```

Comment:
I highly recommend _against_ using submodules. They are a complete PITA for end users.

Reply:
I agree, but this should only affect people wanting to run the tests, I would not consider them users. Cloning each repo during the tests is wasteful, depending on how you do it, and potentially error prune. If we don't want to throw the cloned directories away, it is very difficult to manage it a thread safe manner. Running tests in parallel, either in pytest, tox or nox, would cause issues. I think using submodules would be reasonable here, but I understand if you don't agree.

I can add a check in the tests with message with command to clone submodules. What do you think?

Reply:
> Cloning each repo during the tests is wasteful, depending on how you do it, and potentially error prune.

I personally would prefer doing a shallow copy, .tar.gz download of master branch instead of this. I think that would be more robust. I'd personally would hide it behind a pytest marker, that could be disabled by Linux distributions when validating the package, or they could inject the source package via an environment variable. Submodules are going to be a PITA to maintain.

Reply:
Yeah, I agree about a shallow git clone. I think cloning everything into a tempdir is the right place to start and if we decide that's taking up a lot of resources, we can move it into a private directory at the repo root or something.

That said, I also don't love the idea of a bunch of mocks and a bunch of integration tests as our testing strategies. I think we should just check in a bunch of very simple projects with various different configurations directly.

In the long run, I think that integration testing against real-world projects should probably be handled at a different layer similar to CPython's buildbots. Maybe that's https://github.com/pypa/integration-test, or maybe one day those of us who maintain systematically important projects like this can get our shit together and make something that tests your change against the top N projects on PyPI.